httpbp/client_middlewares.go

```diff
@@ -0,0 +1,196 @@
+package httpbp
+
+import (
+	"io"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	retry "github.com/avast/retry-go"
+	"github.com/opentracing/opentracing-go"
+
+	"github.com/reddit/baseplate.go/retrybp"
+	"github.com/reddit/baseplate.go/tracing"
+	"github.com/sony/gobreaker"
+)
+
+// ClientMiddleware is used to build HTTP client middlewares by implementing
+// http.RoundTripper which http.Client accepts as Transport.
+type ClientMiddleware func(next http.RoundTripper) http.RoundTripper
+
+type roundTripper func(req *http.Request) (*http.Response, error)
+
+func (f roundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	return f(req)
+}
+
+// NewClient returns a standard HTTP client wrapped with the default
+// middlewares plus any additional client middlewares passed into this
+// function.
+func NewClient(config ClientConfig, middlewares ...ClientMiddleware) *http.Client {
+	transport := &http.Transport{
+		MaxConnsPerHost: config.MaxConnections,
+	}
+
+	defaults := []ClientMiddleware{
+		ClientErrorWrapper(),
+		MonitorClient(config),
+		Retries(),
+		MaxConcurrency(config),
```

Comment:
since we cannot add `DrainAndClose` into a middleware, `MaxConcurrency` could be a footgun, so I think we should remove it from default default middlewares.